sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -721,8 +716,55 @@ case class GetExternalRowField(
           "cannot be null.");
       }
 
-      final ${ctx.javaType(dataType)} ${ev.value} = $getField;
+      final Object ${ev.value} = ${row.value}.get($index);
      """
     ev.copy(code = code, isNull = "false")
   }
 }
+
+/**
+ * Validates the actual data type of input expression at runtime.  If it doesn't match the
+ * expectation, throw an exception.
+ */
+case class ValidateExternalType(child: Expression, expected: DataType)
```

Comment:
Previously we completely rely on the correctness of provided schema. If the data doesn't match provided schema, we should see value converting error. I think this `ValidateExternalType` can improve the experience of error handling. However, it looks like a performance regression as we do extra checking here. If we can trust the provided schema, should we do this checking?